crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -1076,6 +1076,111 @@ impl Mesh {
             }
         }
     }
+
+    /// Get a list of this Mesh's triangles if possible.
+    ///
+    /// Returns an error if any of the following conditions are met (see [`MeshTrianglesError`]):
+    /// * The Mesh's [primitive topology] is not `TriangleList` or `TriangleStrip`.
+    /// * The Mesh is missing position or index data.
+    /// * The Mesh's position data has the wrong format (not `Float32x3`).
+    /// * The Mesh has one or more invalid indices.
+    ///
+    /// [primitive topology]: PrimitiveTopology
+    pub fn triangles(&self) -> Result<Vec<Triangle3d>, MeshTrianglesError> {
```

Comment:
We should return an erased `IntoIterator` type here, not a `Vec`. Users can collect on their own here. Normally I wouldn't pick this nit, but meshes are both large and often in the hot loop.

Reply:
Yeah, that's a good point. This turned out to be a bit of a headache since the returned iterator really needs to be a disjunction of four iterators, but I think I got it to work out. Let me know what you think!